torch/_custom_op/impl.py

```diff
@@ -417,6 +415,87 @@ def error_on_ctx():
 
         return inner
 
+    # Should be lazily registered when both backward and save_for_backward
+    # have been provided.
+    def _register_autograd_kernel(self):
+        assert self._has_impl("backward")
+        assert self._has_impl("save_for_backward")
+        kernel = construct_autograd_kernel(
+            self._schema,
+            self._output_differentiability,
+            self,
+            self._get_impl("save_for_backward").func,
+            self._get_impl("backward").func)
+        self._register_impl("autograd", kernel)
+
+    def impl_save_for_backward(self):
+        r"""Register a function that tells us what to save for backward.
+
+        Please see impl_backward for more details.
+        """
+        def inner(f):
+            self._register_impl("save_for_backward", f)
+            if self._has_impl("backward"):
+                self._register_autograd_kernel()
+        return inner
+
+    def impl_backward(self, output_differentiability=None):
+        r"""Registers a backward formula.
+
+        In order for the CustomOp to work with autograd, you need to register
+        a backward formula. There are two pieces to this:
+        1. You must give us a function to specify what to save for backward.
+           Call this the "save for backward" function.
+        2. You must give us a function that computes gradients. Call this the
+           "backward" function.
+
+        Use `impl_save_for_backward` to define a "save for backward" function
+        that specifies what gets saved for backward. The function should accept
+        two arguments ``(inputs, output)`` and return the quantities to be saved
```

Comment:
inputs, outputs ?

Reply:
No, this is actually `inputs, output`:
- If the output is a single Tensor, then `output` is a single Tensor
- If the output is a Tuple of Tensors, then `output` is a Tuple of Tensors
- If the output is a tuple of a single Tensor, then `output` is a Tuple of a single Tensor
I'm open to changing it (though, this is how setup_context works as well).

Reply:
I'm curious:
- Why do you prefer this semantic?
- What is the semantic used in our other similar functions? (hooks always wrap in a tuple IIRC?)

Reply:
> Why do you prefer this semantic?

- It's easier to explain that the `output` is exactly the output of the `forward()`, rather than "we wrap the output in a Tuple if it is not a Tuple".
- It avoids ambiguity: If we support returning TensorList (which we likely will in the future, I have not enabled it in this PR), and we have `foo(Tensor x) -> TensorList`, then there is the question of if `outputs` is a Tuple[List[Tensor]] or just a List[Tensor].

> What is the semantic used in our other similar functions? (hooks always wrap in a tuple IIRC?)

- nn.Module `register_forward_hook` passes the [output as-is without wrapping](https://gist.github.com/zou3519/822505b737f8bc0646e45680bc7798cc)
- `register_backward_hook` (which requires hook of signature `(mod, grad_input, grad_output)` wraps grad_output as a Tuple.
- autograd.Function's forward distinguishes between `return (x,)` and `return x`
- autograd.Function's backward does not distinguish between `return grad` and `return (grad,)`

Based on those examples it seems like for the forward pass we are generally careful that the types match, but when it comes to gradients we generally wrap things in a Tuple to make things simpler.

Reply:
Ok! Keeping it as is sounds good then!